soc/arm/nxp_imx/rt5xx/power.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2022, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/pm/pm.h>
+#include <zephyr/drivers/pinctrl.h>
+#include "fsl_power.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+/*!< Power down all unnecessary blocks */
+#define NODE_ID DT_INST(0, nxp_pdcfg_power)
+#define EXCLUDE_FROM_DEEPSLEEP ((const uint32_t[]) \
+					DT_PROP_OR(NODE_ID, deep_sleep_config, {}))
+
+#define EXCLUDE_FROM_DEEP_POWERDOWN ((const uint32_t[]){0, 0, 0, 0})
+
+/* System clock frequency. */
+extern uint32_t SystemCoreClock;
+
+#define FLEXSPI_DEVICE DT_NODELABEL(flexspi)
+#if DT_NODE_HAS_STATUS(FLEXSPI_DEVICE, okay)
+
+/*
+ * Declare pin configuration state for flexspi pins.
+ * These pins will be reconfigured for power savings
+ */
+PINCTRL_DT_DEFINE(FLEXSPI_DEVICE);
+const struct pinctrl_dev_config *flexspi_pincfg = PINCTRL_DT_DEV_CONFIG_GET(FLEXSPI_DEVICE);
+const struct pinctrl_state *flexspi_pins_default_state;
+const struct pinctrl_state *flexspi_pins_sleep_state;
+/* Flag indicates if we can reconfigure flexspi pins for additional power savings */
+static bool flexspi_pin_reconfig;
+
+#endif
+
+#define FLEXSPI2_DEVICE DT_NODELABEL(flexspi2)
+#if DT_NODE_HAS_STATUS(FLEXSPI2_DEVICE, okay)
+
+/*
+ * Declare pin configuration state for flexspi2 pins.
+ * These pins will be reconfigured for power savings
+ */
+PINCTRL_DT_DEFINE(FLEXSPI2_DEVICE);
+const struct pinctrl_dev_config *flexspi2_pincfg = PINCTRL_DT_DEV_CONFIG_GET(FLEXSPI2_DEVICE);
+const struct pinctrl_state *flexspi2_pins_default_state;
+const struct pinctrl_state *flexspi2_pins_sleep_state;
+/* Flag indicates if we can reconfigure flexspi pins for additional power savings */
+static bool flexspi2_pin_reconfig;
```

Comment:
this needs to be managed by the device, not outside. Don't you have a driver for flexspi2?

Reply:
These need to be set close to entering Low Power modes. Hence they are done here.

Reply:
The problem with this approach is that it will conflict with e.g. dynamic pinctrl (you'll end up with duplicate symbols). What I'd suggest is to create a new compatible where to store such description (maybe re-use nxp,pdcfg-power by adding there pinctrl-N entries). Also, please document that this has to be done independently of the flash driver to save power in all configurations.

Reply:
@gmarull, thank you for your comments. Turns out we can enable the FlexSPI driver without having to pull in the whole Flash subsystem.
I have moved this code to the FlexSPI driver by adding Device Level Power Management. This did uncover an issue in the Zephyr PM code. I will submit a separate PR for that.

Reply:
@gmarull, let me know if you have any more comments. We would like to get this is in time for the next release.